<a href="https://colab.research.google.com/github/eridanletalis/linalg/blob/master/linalg_hw_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$




In [3]:
A = np.array([[1, 2, 0],
              [0,0,5],
              [3, -4, 2],
              [1,6,5],
              [0,1,0]])




Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [4]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [8]:
print(f'Матрица D:\n{D}')
print(f'Матрица U:\n{U}')
print(f'Матрица V:\n{V}')

print(f"Выполним проверку: \n {np.dot(np.dot(U, D), V.T)}")
print(f"Исходная матрица A:\n{A}")

Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]
Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]
Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]
Выполним проверку: 
 [[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1. -0.]]
Исходная матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

In [15]:
D[0,0]

8.824868854820444

По определению,  евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел $D$. Максимальное значение отношения $\left (\frac{(Dz,Dz)}{(z,z)}\right )$ будет равно максимальному числу $\mu_{max}$. Так как в диагональной матрице $D$ элементы отсортированы по убыванию от $mu_{1}$ до $mu_{n}$, то евклидовой нормой будет число $mu_{1}$.

$$
\left \| A \right \|_{E}=\mu_{1} = 8.82.
$$

Норма Фробениуса по сингулярному разложению вычисляется так:
$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

Что программно реализуется ниже

In [30]:
norm = np.sqrt(np.array([D[i,i] **2  for i in range(0, np.linalg.matrix_rank(D))]).sum())
print(f"Норма Фробениуса равна: {norm}")

Норма Фробениуса равна: 11.045361017187263


In [31]:
np.linalg.norm(A)

11.045361017187261